agent/config/config.go

```diff
@@ -359,6 +361,40 @@ func fileConfig() (Config, error) {
 	return cfg, nil
 }
 
+// userDataConfig reads configuration JSON from instance's userdata. It doesn't
+// return any error as it's entirely optional to configure the ECS agent using
+// this method.
+// Example:
+// {"ECSAgentConfiguration":{"Cluster":"default"}}
+func userDataConfig(ec2Client ec2.EC2MetadataClient) Config {
+	type userDataParser struct {
+		Config Config `json:"ECSAgentConfiguration"`
+	}
+
+	parsedUserData := userDataParser{
+		Config: Config{},
+	}
+
+	userData, err := ec2Client.GetUserData()
+	if err != nil {
+		seelog.Warnf("Unable to fetch user data: %v", err)
+		// Unable to read userdata from instance metadata. Just
+		// return early
+		return parsedUserData.Config
+	}
+	// In the future, if we want to support base64 encoded config,
+	// we'd need to add logic to decode the string here.
+	err = json.Unmarshal([]byte(userData), &parsedUserData)
+	if err != nil {
+		seelog.Warnf("Unable to parse user data: %v", err)
+		// Unable to parse userdata as a valid JSON. Return the
+		// empty config
+		return parsedUserData.Config
```

Comment:
Are we sure this would be empty `Config` as opposed to some junk value? 

Imo, it might be better to return empty Config object.

Reply:
agreed, making this change.